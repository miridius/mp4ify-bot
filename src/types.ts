import type { Context } from 'telegraf';
import type { Message, Update } from 'telegraf/types';

export type MessageContext = Context<Update.MessageUpdate<Message.TextMessage>>;
export type InlineQueryContext = Context<Update.InlineQueryUpdate>;

export type AnyContext = MessageContext | InlineQueryContext;
