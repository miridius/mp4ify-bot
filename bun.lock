{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "/app",
      "dependencies": {
        "telegraf": "^4.16.3",
      },
      "devDependencies": {
        "@types/bun": "latest",
        "oxlint": "^0.16.10",
        "prettier": "^3.5.3",
        "simple-git-hooks": "^2.13.0",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@oxlint/darwin-arm64": ["@oxlint/darwin-arm64@0.16.10", "", { "os": "darwin", "cpu": "arm64" }, "sha512-lNnkf7/auNXXD7gU3IbWMyuYiKj4Va+hhoe6+NwGfAjpKm5L4Uvc93o7tay163WVzQePVUOz/IwsJPLQs3hcjQ=="],

    "@oxlint/darwin-x64": ["@oxlint/darwin-x64@0.16.10", "", { "os": "darwin", "cpu": "x64" }, "sha512-axqbGX//v+AF/73OZUfUGG5pZidnZf9T+wc+1Jf19O0DViPpcmbVH7l0aLImFQuRn6Nl9LAbfBbKAgO1BS+OSg=="],

    "@oxlint/linux-arm64-gnu": ["@oxlint/linux-arm64-gnu@0.16.10", "", { "os": "linux", "cpu": "arm64" }, "sha512-CoFYLjmttdBRw6k3waRuSoW+36ojY8WspOkr5KzNM6GytrKKIRSLUNkaLLSSnXQK6MN13jemJa32U1DRn8/V1Q=="],

    "@oxlint/linux-arm64-musl": ["@oxlint/linux-arm64-musl@0.16.10", "", { "os": "linux", "cpu": "arm64" }, "sha512-jidO59qo6RZ1HRk7a27noIgUeJmzju13NcKm7+5zZObPPGuXXO1OT3ZYabyO4Zz3eU+wqMTgjM4yphbJerkxPQ=="],

    "@oxlint/linux-x64-gnu": ["@oxlint/linux-x64-gnu@0.16.10", "", { "os": "linux", "cpu": "x64" }, "sha512-RWbkaDIX4fD+OaE+XfNG8nOxh+qLA90KhuejlTFbAApn7lZtyhr/UTIPuJsjmzMUW145C0NJTugOB6Q6XtKGNA=="],

    "@oxlint/linux-x64-musl": ["@oxlint/linux-x64-musl@0.16.10", "", { "os": "linux", "cpu": "x64" }, "sha512-V7oXFLCowIEOnYBph4J2KSiHYtBIrx00xmwzKzSvwMzVOxDkw1LmPUYLCz/AEU6MctE6S2ZR947IXJY7ch4HeQ=="],

    "@oxlint/win32-arm64": ["@oxlint/win32-arm64@0.16.10", "", { "os": "win32", "cpu": "arm64" }, "sha512-QrWe71tjRKSDLejklPoU7gySq16nrJMOrqcSaw+ERQtJpnJBruF0RWQEVFEev6TvzwRnxrdTqDucYcFmoFqGeg=="],

    "@oxlint/win32-x64": ["@oxlint/win32-x64@0.16.10", "", { "os": "win32", "cpu": "x64" }, "sha512-V5LHfgB5uiiKNpOFOAfygsL36pN6qSuQ0moiuEyZFfWL2Iyx5026ksefiD9nXcygQsLYU7f6xQbOINAY3uIvcw=="],

    "@telegraf/types": ["@telegraf/types@7.1.0", "", {}, "sha512-kGevOIbpMcIlCDeorKGpwZmdH7kHbqlk/Yj6dEpJMKEQw5lk0KVQY0OLXaCswy8GqlIVLd5625OB+rAntP9xVw=="],

    "@types/bun": ["@types/bun@1.2.11", "", { "dependencies": { "bun-types": "1.2.11" } }, "sha512-ZLbbI91EmmGwlWTRWuV6J19IUiUC5YQ3TCEuSHI3usIP75kuoA8/0PVF+LTrbEnVc8JIhpElWOxv1ocI1fJBbw=="],

    "@types/node": ["@types/node@22.15.3", "", { "dependencies": { "undici-types": "~6.21.0" } }, "sha512-lX7HFZeHf4QG/J7tBZqrCAXwz9J5RD56Y6MpP0eJkka8p+K0RY/yBTW7CYFJ4VGCclxqOLKmiGP5juQc6MKgcw=="],

    "abort-controller": ["abort-controller@3.0.0", "", { "dependencies": { "event-target-shim": "^5.0.0" } }, "sha512-h8lQ8tacZYnR3vNQTgibj+tODHI5/+l06Au2Pcriv/Gmet0eaj4TwWH41sO9wnHDiQsEj19q0drzdWdeAHtweg=="],

    "buffer-alloc": ["buffer-alloc@1.2.0", "", { "dependencies": { "buffer-alloc-unsafe": "^1.1.0", "buffer-fill": "^1.0.0" } }, "sha512-CFsHQgjtW1UChdXgbyJGtnm+O/uLQeZdtbDo8mfUgYXCHSM1wgrVxXm6bSyrUuErEb+4sYVGCzASBRot7zyrow=="],

    "buffer-alloc-unsafe": ["buffer-alloc-unsafe@1.1.0", "", {}, "sha512-TEM2iMIEQdJ2yjPJoSIsldnleVaAk1oW3DBVUykyOLsEsFmEc9kn+SFFPz+gl54KQNxlDnAwCXosOS9Okx2xAg=="],

    "buffer-fill": ["buffer-fill@1.0.0", "", {}, "sha512-T7zexNBwiiaCOGDg9xNX9PBmjrubblRkENuptryuI64URkXDFum9il/JGL8Lm8wYfAXpredVXXZz7eMHilimiQ=="],

    "bun-types": ["bun-types@1.2.11", "", { "dependencies": { "@types/node": "*" } }, "sha512-dbkp5Lo8HDrXkLrONm6bk+yiiYQSntvFUzQp0v3pzTAsXk6FtgVMjdQ+lzFNVAmQFUkPQZ3WMZqH5tTo+Dp/IA=="],

    "debug": ["debug@4.4.0", "", { "dependencies": { "ms": "^2.1.3" } }, "sha512-6WTZ/IxCY/T6BALoZHaE4ctp9xm+Z5kY/pzYaCHRFeyVhojxlrm+46y68HA6hr0TcwEssoxNiDEUJQjfPZ/RYA=="],

    "event-target-shim": ["event-target-shim@5.0.1", "", {}, "sha512-i/2XbnSz/uxRCU6+NdVJgKWDTM427+MqYbkQzD321DuCQJUqOuJKIA0IM2+W2xtYHdKOmZ4dR6fExsd4SXL+WQ=="],

    "mri": ["mri@1.2.0", "", {}, "sha512-tzzskb3bG8LvYGFF/mDTpq3jpI6Q9wc3LEmBaghu+DdCssd1FakN7Bc0hVNmEyGq1bq3RgfkCb3cmQLpNPOroA=="],

    "ms": ["ms@2.1.3", "", {}, "sha512-6FlzubTLZG3J2a/NVCAleEhjzq5oxgHyaCU9yYXvcLsvoVaHJq/s5xXI6/XXP6tz7R9xAOtHnSO/tXtF3WRTlA=="],

    "node-fetch": ["node-fetch@2.7.0", "", { "dependencies": { "whatwg-url": "^5.0.0" }, "peerDependencies": { "encoding": "^0.1.0" }, "optionalPeers": ["encoding"] }, "sha512-c4FRfUm/dbcWZ7U+1Wq0AwCyFL+3nt2bEw05wfxSz+DWpWsitgmSgYmy2dQdWyKC1694ELPqMs/YzUSNozLt8A=="],

    "oxlint": ["oxlint@0.16.10", "", { "optionalDependencies": { "@oxlint/darwin-arm64": "0.16.10", "@oxlint/darwin-x64": "0.16.10", "@oxlint/linux-arm64-gnu": "0.16.10", "@oxlint/linux-arm64-musl": "0.16.10", "@oxlint/linux-x64-gnu": "0.16.10", "@oxlint/linux-x64-musl": "0.16.10", "@oxlint/win32-arm64": "0.16.10", "@oxlint/win32-x64": "0.16.10" }, "bin": { "oxlint": "bin/oxlint", "oxc_language_server": "bin/oxc_language_server" } }, "sha512-xSHmXzQ8cuoeUFoHsc0NIab81KsXNNUKJbK7ZaoCcSWcWSDc4oyc9giPnaVOnVCye5XR7FfX2x4OXqJ9nQdKHw=="],

    "p-timeout": ["p-timeout@4.1.0", "", {}, "sha512-+/wmHtzJuWii1sXn3HCuH/FTwGhrp4tmJTxSKJbfS+vkipci6osxXM5mY0jUiRzWKMTgUT8l7HFbeSwZAynqHw=="],

    "prettier": ["prettier@3.5.3", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-QQtaxnoDJeAkDvDKWCLiwIXkTgRhwYDEQCghU9Z6q03iyek/rxRh/2lC3HB7P8sWT2xC/y5JDctPLBIGzHKbhw=="],

    "safe-compare": ["safe-compare@1.1.4", "", { "dependencies": { "buffer-alloc": "^1.2.0" } }, "sha512-b9wZ986HHCo/HbKrRpBJb2kqXMK9CEWIE1egeEvZsYn69ay3kdfl9nG3RyOcR+jInTDf7a86WQ1d4VJX7goSSQ=="],

    "sandwich-stream": ["sandwich-stream@2.0.2", "", {}, "sha512-jLYV0DORrzY3xaz/S9ydJL6Iz7essZeAfnAavsJ+zsJGZ1MOnsS52yRjU3uF3pJa/lla7+wisp//fxOwOH8SKQ=="],

    "simple-git-hooks": ["simple-git-hooks@2.13.0", "", { "bin": { "simple-git-hooks": "cli.js" } }, "sha512-N+goiLxlkHJlyaYEglFypzVNMaNplPAk5syu0+OPp/Bk6dwVoXF6FfOw2vO0Dp+JHsBaI+w6cm8TnFl2Hw6tDA=="],

    "telegraf": ["telegraf@4.16.3", "", { "dependencies": { "@telegraf/types": "^7.1.0", "abort-controller": "^3.0.0", "debug": "^4.3.4", "mri": "^1.2.0", "node-fetch": "^2.7.0", "p-timeout": "^4.1.0", "safe-compare": "^1.1.4", "sandwich-stream": "^2.0.2" }, "bin": { "telegraf": "lib/cli.mjs" } }, "sha512-yjEu2NwkHlXu0OARWoNhJlIjX09dRktiMQFsM678BAH/PEPVwctzL67+tvXqLCRQQvm3SDtki2saGO9hLlz68w=="],

    "tr46": ["tr46@0.0.3", "", {}, "sha512-N3WMsuqV66lT30CrXNbEjx4GEwlow3v6rr4mCcv6prnfwhS01rkgyFdjPNBYd9br7LpXV1+Emh01fHnq2Gdgrw=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "undici-types": ["undici-types@6.21.0", "", {}, "sha512-iwDZqg0QAGrg9Rav5H4n0M64c3mkR59cJ6wQp+7C4nI0gsmExaedaYLNO44eT4AtBBwjbTiGPMlt2Md0T9H9JQ=="],

    "webidl-conversions": ["webidl-conversions@3.0.1", "", {}, "sha512-2JAn3z8AR6rjK8Sm8orRC0h/bcl/DqL7tRPdGZ4I1CjdF+EaMLmYxBHyXuKL849eucPFhvBoxMsflfOb8kxaeQ=="],

    "whatwg-url": ["whatwg-url@5.0.0", "", { "dependencies": { "tr46": "~0.0.3", "webidl-conversions": "^3.0.0" } }, "sha512-saE57nupxk6v3HY35+jzBwYa0rKSy0XR8JSxZPwgLr7ys0IBzhGviA1/TUGJLmSVqs8pb9AnvICXEuOHLprYTw=="],
  }
}
