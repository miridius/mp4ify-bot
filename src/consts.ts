import { $ } from 'bun';
import { randomBytes } from 'node:crypto';

export const hostname = (await $`hostname`.text()).trim(); // getRequiredEnv('HOSTNAME');

export const secret = randomBytes(32).toString('hex');

const getRequiredEnv = (name: string) => {
  const value = Bun.env[name];
  if (!value) throw new Error(`${name} env var is required`);
  return value;
};
``;
export const botToken = getRequiredEnv('BOT_TOKEN');

export const apiRoot = Bun.env.API_ROOT || 'http://bot-api:8081';
