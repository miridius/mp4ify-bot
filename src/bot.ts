import { Telegraf } from 'telegraf';
import { editedMessage, message } from 'telegraf/filters';
import { apiRoot } from './consts';
import { inlineQueryHandler, textMessageHandler } from './handlers';

export const start = async (botToken: string) => {
  const bot = new Telegraf(botToken, { telegram: { apiRoot } });
  console.debug(bot.telegram.options);

  bot.on(message('text'), (ctx) => textMessageHandler(ctx));
  bot.on(editedMessage('text'), (ctx) => textMessageHandler(ctx));
  bot.on('inline_query', (ctx) => inlineQueryHandler(ctx));

  bot.use((ctx) => console.log('unhandled update:', ctx.update));

  bot.launch();
  // wait for the bot to start
  while (!(bot as any).polling) await Bun.sleep(100);

  // Enable graceful stop
  process.once('SIGINT', () => bot.stop('SIGINT'));
  process.once('SIGTERM', () => bot.stop('SIGTERM'));

  return bot;
};
