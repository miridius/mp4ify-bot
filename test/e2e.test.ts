import { $ } from 'bun';
import {
  afterAll,
  beforeEach,
  describe,
  expect,
  it,
  jest,
  mock,
} from 'bun:test';
import { downloadVideo, getInfo, sendVideo } from '../src/download-video';
import { withBotApi } from './simulate-bot-api';
import { spyMock, waitUntil } from './test-utils';

beforeEach(() => jest.clearAllMocks());
afterAll(() => mock.restore());
spyMock(console, 'debug');
spyMock(console, 'table');

// e2e tests:
// 1. url each from youtube, insta, reddit
// 2. use real yt-dlp, but use fixtures for calls to bot-api
// 3. test message and in-line
// 4. test video id cache

const hiMessage = { text: 'hi' };

const urlMessage = (url: string, verbose?: boolean) => ({
  text: verbose ? `/verbose ${url}` : url,
  entities: [
    {
      offset: verbose ? 9 : 0,
      length: (verbose ? 9 : 0) + url.length,
      type: 'url' as const,
    },
  ],
  link_preview_options: { is_disabled: true },
});

const testUrls = [
  'https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==',
  // alias url
  'https://www.reddit.com/r/nextfuckinglevel/s/iGEii0a7V6',
  // canonical url for same video
  'https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo',
  // fails if we run the test too often
  'http://youtube.com/shorts/0COu-qMC18Y',
];

const clearDiskCache = async () => $`rm -rf /storage/*`.catch(() => null);

const clearInMemoryCache = () => {
  getInfo.cache.clear();
  downloadVideo.cache.clear();
  sendVideo.cache.clear();
};

describe.if(!!Bun.env.TEST_E2E)('message handler', async () => {
  await clearDiskCache();
  clearInMemoryCache();

  it('ignores messages without urls', () =>
    withBotApi(async (api) => {
      api.sendTextMessageToBot(hiMessage);
      await Bun.sleep(1000);
      expect(api.sentMessages).toMatchInlineSnapshot(`[]`);
    }));

  it.each(testUrls)(
    'downloads %s',
    (url) =>
      withBotApi(async (api) => {
        const waitForVideo = (ms: number) =>
          waitUntil(
            () =>
              api.sentMessages.length > 1 ||
              api.sentMessages.some(({ text }) =>
                text?.includes('ðŸ’¥ <b>Download failed</b>:'),
              ),
            ms,
          );

        // initial download
        clearInMemoryCache();
        api.sendTextMessageToBot(urlMessage(url));
        await waitForVideo(25_000);
        expect(api.sentMessages).toMatchSnapshot('download');

        // in memory cache
        api.sentMessages.length = 0;
        api.sendTextMessageToBot(urlMessage(url));
        await waitForVideo(5_000);
        expect(api.sentMessages).toMatchSnapshot('mem cache');

        // disk cache
        clearInMemoryCache();
        api.sentMessages.length = 0;
        api.sendTextMessageToBot(urlMessage(url));
        await waitForVideo(5_000);
        expect(api.sentMessages).toMatchSnapshot('disk cache');
      }),
    40_000,
  );
});

describe.todo('inline query handler');
