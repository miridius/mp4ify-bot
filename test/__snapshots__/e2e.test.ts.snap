// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`message handler downloads https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==: download 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 0,
    },
    "text": 
"üßê <b>Scraping</b> https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==...

üé¨ <b>Video info:</b>

<b>URL</b>: https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==
<b>filename</b>: Video_by_disastra_memes-[DKbYQgeoL3F].3.mp4
<b>duration</b>: 12 sec
<b>resolution</b>: 640x584

‚¨áÔ∏è <b>Downloading...</b>

üöÄ <b>Uploading (0.41 MB)...</b>"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 12,
    "height": 584,
    "reply_parameters": {
      "message_id": 0,
    },
    "reply_to_message_id": 0,
    "supports_streaming": true,
    "video": "file:///storage/Instagram/Video_by_disastra_memes-%5BDKbYQgeoL3F%5D.3.mp4",
    "width": 640,
  },
]
`;

exports[`message handler downloads https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==: mem cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 1,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==
<b>filename</b>: Video_by_disastra_memes-[DKbYQgeoL3F].3.mp4
<b>duration</b>: 12 sec
<b>resolution</b>: 640x584"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 12,
    "height": 584,
    "reply_parameters": {
      "message_id": 1,
    },
    "reply_to_message_id": 1,
    "supports_streaming": true,
    "video": "11ej1fdh590e0",
    "width": 640,
  },
]
`;

exports[`message handler downloads https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==: disk cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 2,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.instagram.com/reel/DKbYQgeoL3F/?igsh=MTh4MnpnYm9hdjJ5OA==
<b>filename</b>: Video_by_disastra_memes-[DKbYQgeoL3F].3.mp4
<b>duration</b>: 12 sec
<b>resolution</b>: 640x584"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 12,
    "height": 584,
    "reply_parameters": {
      "message_id": 2,
    },
    "reply_to_message_id": 2,
    "supports_streaming": true,
    "video": "11ej1fdh590e0",
    "width": 640,
  },
]
`;

exports[`message handler downloads https://www.reddit.com/r/nextfuckinglevel/s/iGEii0a7V6: download 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 0,
    },
    "text": 
"üßê <b>Scraping</b> https://www.reddit.com/r/nextfuckinglevel/s/iGEii0a7V6...

üé¨ <b>Video info:</b>

<b>URL</b>: https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo&utm_content=2&utm_medium=android_app&utm_name=androidcss&utm_source=share&utm_term=1
<b>filename</b>: Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-[auhgg4yoeo5f1].hls-1023+dash-6.mp4
<b>duration</b>: 10 sec
<b>size</b>: 0.01 MB
<b>resolution</b>: 480x600
<b>video codec</b>: avc1.4D401F @ 1023.275 kbps
<b>audio codec</b>: mp4a.40.2 @ 132.978 kbps

‚¨áÔ∏è <b>Downloading...</b>

üöÄ <b>Uploading (1.20 MB)...</b>"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 10,
    "height": 600,
    "reply_parameters": {
      "message_id": 0,
    },
    "reply_to_message_id": 0,
    "supports_streaming": true,
    "video": "file:///storage/Reddit/Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-%5Bauhgg4yoeo5f1%5D.hls-1023+dash-6.mp4",
    "width": 480,
  },
]
`;

exports[`message handler downloads https://www.reddit.com/r/nextfuckinglevel/s/iGEii0a7V6: mem cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 1,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo&utm_content=2&utm_medium=android_app&utm_name=androidcss&utm_source=share&utm_term=1
<b>filename</b>: Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-[auhgg4yoeo5f1].hls-1023+dash-6.mp4
<b>duration</b>: 10 sec
<b>size</b>: 0.01 MB
<b>resolution</b>: 480x600
<b>video codec</b>: avc1.4D401F @ 1023.275 kbps
<b>audio codec</b>: mp4a.40.2 @ 132.978 kbps"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 10,
    "height": 600,
    "reply_parameters": {
      "message_id": 1,
    },
    "reply_to_message_id": 1,
    "supports_streaming": true,
    "video": "uufngxsbt7v5",
    "width": 480,
  },
]
`;

exports[`message handler downloads https://www.reddit.com/r/nextfuckinglevel/s/iGEii0a7V6: disk cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 2,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo&utm_content=2&utm_medium=android_app&utm_name=androidcss&utm_source=share&utm_term=1
<b>filename</b>: Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-[auhgg4yoeo5f1].hls-1023+dash-6.mp4
<b>duration</b>: 10 sec
<b>size</b>: 0.01 MB
<b>resolution</b>: 480x600
<b>video codec</b>: avc1.4D401F @ 1023.275 kbps
<b>audio codec</b>: mp4a.40.2 @ 132.978 kbps"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 10,
    "height": 600,
    "reply_parameters": {
      "message_id": 2,
    },
    "reply_to_message_id": 2,
    "supports_streaming": true,
    "video": "uufngxsbt7v5",
    "width": 480,
  },
]
`;

exports[`message handler downloads https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo: download 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 0,
    },
    "text": 
"üßê <b>Scraping</b> https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo...

üé¨ <b>Video info:</b>

<b>URL</b>: https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo
<b>filename</b>: Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-[auhgg4yoeo5f1].hls-1023+dash-6.mp4
<b>duration</b>: 10 sec
<b>size</b>: 0.01 MB
<b>resolution</b>: 480x600
<b>video codec</b>: avc1.4D401F @ 1023.275 kbps
<b>audio codec</b>: mp4a.40.2 @ 132.978 kbps

‚¨áÔ∏è <b>Downloading...</b>

üöÄ <b>Uploading (1.20 MB)...</b>"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 10,
    "height": 600,
    "reply_parameters": {
      "message_id": 0,
    },
    "reply_to_message_id": 0,
    "supports_streaming": true,
    "video": "file:///storage/Reddit/Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-%5Bauhgg4yoeo5f1%5D.hls-1023+dash-6.mp4",
    "width": 480,
  },
]
`;

exports[`message handler downloads https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo: mem cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 1,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo
<b>filename</b>: Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-[auhgg4yoeo5f1].hls-1023+dash-6.mp4
<b>duration</b>: 10 sec
<b>size</b>: 0.01 MB
<b>resolution</b>: 480x600
<b>video codec</b>: avc1.4D401F @ 1023.275 kbps
<b>audio codec</b>: mp4a.40.2 @ 132.978 kbps"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 10,
    "height": 600,
    "reply_parameters": {
      "message_id": 1,
    },
    "reply_to_message_id": 1,
    "supports_streaming": true,
    "video": "uufngxsbt7v5",
    "width": 480,
  },
]
`;

exports[`message handler downloads https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo: disk cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 2,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.reddit.com/r/nextfuckinglevel/comments/1l68isw/mix_of_coolness_agility_technique_power_and_a/?share_id=ejTJZnh_f4BZuzlnfcOUo
<b>filename</b>: Mix_of_coolness_agility_technique_power_and_a_touch_of_madness-[auhgg4yoeo5f1].hls-1023+dash-6.mp4
<b>duration</b>: 10 sec
<b>size</b>: 0.01 MB
<b>resolution</b>: 480x600
<b>video codec</b>: avc1.4D401F @ 1023.275 kbps
<b>audio codec</b>: mp4a.40.2 @ 132.978 kbps"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 10,
    "height": 600,
    "reply_parameters": {
      "message_id": 2,
    },
    "reply_to_message_id": 2,
    "supports_streaming": true,
    "video": "uufngxsbt7v5",
    "width": 480,
  },
]
`;

exports[`message handler downloads http://youtube.com/shorts/0COu-qMC18Y: download 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 0,
    },
    "text": 
"üßê <b>Scraping</b> http://youtube.com/shorts/0COu-qMC18Y...

üé¨ <b>Video info:</b>

<b>URL</b>: https://www.youtube.com/watch?v=0COu-qMC18Y
<b>filename</b>: Zinger_Burgers-[0COu-qMC18Y].299+140.mp4
<b>duration</b>: 57 sec
<b>size</b>: 24.93 MB
<b>resolution</b>: 1080x1920
<b>video codec</b>: avc1.64002a @ 3571.925 kbps
<b>audio codec</b>: mp4a.40.2 @ 129.585 kbps

‚¨áÔ∏è <b>Downloading...</b>

üöÄ <b>Uploading (24.95 MB)...</b>"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 57,
    "height": 1920,
    "reply_parameters": {
      "message_id": 0,
    },
    "reply_to_message_id": 0,
    "supports_streaming": true,
    "video": "file:///storage/youtube/Zinger_Burgers-%5B0COu-qMC18Y%5D.299+140.mp4",
    "width": 1080,
  },
]
`;

exports[`message handler downloads http://youtube.com/shorts/0COu-qMC18Y: mem cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 1,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.youtube.com/watch?v=0COu-qMC18Y
<b>filename</b>: Zinger_Burgers-[0COu-qMC18Y].299+140.mp4
<b>duration</b>: 57 sec
<b>size</b>: 24.93 MB
<b>resolution</b>: 1080x1920
<b>video codec</b>: avc1.64002a @ 3571.925 kbps
<b>audio codec</b>: mp4a.40.2 @ 129.585 kbps"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 57,
    "height": 1920,
    "reply_parameters": {
      "message_id": 1,
    },
    "reply_to_message_id": 1,
    "supports_streaming": true,
    "video": "7jef1dex1i0d",
    "width": 1080,
  },
]
`;

exports[`message handler downloads http://youtube.com/shorts/0COu-qMC18Y: disk cache 1`] = `
[
  {
    "chat_id": 1337,
    "disable_notification": true,
    "link_preview_options": {
      "is_disabled": true,
    },
    "parse_mode": "HTML",
    "reply_parameters": {
      "message_id": 2,
    },
    "text": 
"
üé¨ <b>Video info:</b>

<b>URL</b>: https://www.youtube.com/watch?v=0COu-qMC18Y
<b>filename</b>: Zinger_Burgers-[0COu-qMC18Y].299+140.mp4
<b>duration</b>: 57 sec
<b>size</b>: 24.93 MB
<b>resolution</b>: 1080x1920
<b>video codec</b>: avc1.64002a @ 3571.925 kbps
<b>audio codec</b>: mp4a.40.2 @ 129.585 kbps"
,
  },
  {
    "chat_id": 1337,
    "disable_notification": true,
    "duration": 57,
    "height": 1920,
    "reply_parameters": {
      "message_id": 2,
    },
    "reply_to_message_id": 2,
    "supports_streaming": true,
    "video": "7jef1dex1i0d",
    "width": 1080,
  },
]
`;
